src/v/kafka/server/handlers/metadata.cc

```diff
@@ -422,4 +423,31 @@ ss::future<response_ptr> metadata_handler::handle(
     co_return co_await ctx.respond(std::move(reply));
 }
 
+size_t metadata_memory_estimator(size_t) {
+    // We cannot make a precise estimate of the size of a metadata response by
+    // examining only the size of the request (nor even by examining the entire
+    // request) since the response depends on the number of partitions in the
+    // cluster. Instead, we return a conservative estimate based on the soft
```

Comment:
Does it require much more plumbing to also implement the return of un-needed units once the request is read?  It's not essential for robustness, but i'd like to have confidence that the code structure here is sufficient for adding that.

Reply:
There are a few places that we can release units, but I think "mostly yes".

Specifically, the main place we can release units is when the response processing has been been completed (the handler is done) and we just need to finish writing out the request: at this point we know the size of the request is simply the encoded size of the response (plus perhaps some padding due to allocation buckets, though we hardly handle this anywhere currently) so we can reduce the units to that amount.

This can be done generically w/o help from the handler (and in fact is quite easy, I just omitted it here out of an abundance of caution given that we'd like to get this into 22.2).

The other place would be inside the request processing, e.g., at certain points where it it is known that large intermediate structures have been cleaned up, we can free units. This can be part of the dynamic two-phase allocation as described above and would be up to each handler. To do this, we need to pass the units into somewhere the handlers can access them, e.g., the request context, which is where I was heading with much of this plumbing. I don't actually plumb it into the request context _yet_ as it is not used in this stopgap fix.